trace/opencensus/trace/ext/grpc/client_interceptor.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2017, OpenCensus Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import grpc
+import six
+import sys
+
+from opencensus.trace import grpc as oc_grpc
+from opencensus.trace.grpc import grpc_ext
+from opencensus.trace.propagation import binary_format
+from opencensus.trace import execution_context
+
+from opencensus.trace.enums import Enum
+
+log = logging.getLogger(__name__)
+
+
+class OpenCensusClientInterceptor(grpc_ext.UnaryUnaryClientInterceptor,
+                                  grpc_ext.UnaryStreamClientInterceptor,
+                                  grpc_ext.StreamUnaryClientInterceptor,
+                                  grpc_ext.StreamStreamClientInterceptor):
+
+    def __init__(self, tracer=None):
+        if tracer is None:
+            tracer = execution_context.get_opencensus_tracer()
+
+        self._tracer = tracer
+        self._propagator = binary_format.BinaryFormatPropagator()
+
+    def _start_client_span(self, method):
+        log.info('Start client span')
+        span = self._tracer.start_span(name='[gRPC]{}'.format(str(method)))
+        span.add_label(label_key='component', label_value='grpc')
+        span.kind = Enum.SpanKind.RPC_CLIENT
+        return span
+
+    def _trace_async_result(self, result):
+        if isinstance(result, grpc.Future):
+            result.add_done_callback(self._future_done_callback())
+
+        return result
+
+    def _future_done_callback(self):
+        def callback(future_response):
+            code = future_response.code()
+
+            if code != grpc.StatusCode.OK:
+                span.add_label('error in response', str(code))
+
+            response = future_response.result()
+            self._tracer.end_span()
+
+        return callback
+
+    def intercept_call(self, request_type, invoker, method, request, **kwargs):
+        metadata = getattr(kwargs, 'metadata', ())
+
+        with self._start_client_span(method) as span:
+            span_context = self._tracer.span_context
+            header = self._propagator.to_header(span_context)
+            grpc_trace_metadata = {
+                oc_grpc.GRPC_TRACE_KEY: header,
+            }
+            metadata = metadata + tuple(six.iteritems(grpc_trace_metadata))
+
+            span.name = '[gRPC_client][{}]{}'.format(request_type, str(method))
```

Comment:
As per our discussion, the hostname/port is part of the grpc Channel object, and we should probably record that.  I'm not sure how we dig out the channel from the interceptor, we might have to change `intercept_channel`.

Reply:
Actually, I'm not sure where the hostname/port is stored, looking at the grpc docs.

Reply:
Just confirmed with Mehrdad, we cannot access the host:port in the interceptors since the interceptors are operating at the front-end of the RPC stack, and RPCs are dispatched at lower level. But we can let the user pass in the host:port when initialize the `OpenCensusClientInterceptor`, and then add this information to span labels.